### PR TITLE
Move detect_areca in the areca module

### PR DESCRIPTION
**Pull request by elfosardo**
Start moving stuff out of detect.
That will also allow us to stop passing the same huge list of data
between functions and modules.
In the long term it will help testing and converting to a more sane
schema.

### CODE REVIEW

**Comment by elfosardo** [comment]
we should give priority to #138  before anything else

**Comment by elfosardo** [comment]
@dtantsur @tbreeds @ErwanAliasr1  wondering if we should cut a release before merging this ?

**Comment by ErwanAliasr1** [comment]
> @dtantsur @tbreeds @ErwanAliasr1 wondering if we should cut a release before merging this ?

Good idea. Just pushed 0.25.0 !

File: hardware/detect.py [hunk 1]
```diff
@@ -990,7 +944,9 @@ def main():
     args = parse_args(sys.argv[1:])
 
     hrdw = []
-    detect_areca(hrdw)
+    areca_hrdw = areca.detect()
```
Comment (ErwanAliasr1) [inline hardware/detect.py:947]: I like the change, this is clearly a good idea for the other modules too.

That said I'm not a big fan of having to process an if areca_hrdw after the call. Can we keep a similar way to add elements into the hrdw structure ?

Comment (elfosardo) [reply hardware/detect.py:947]: The idea is exactly give a base for a change in the other modules too, so I'm glad it got some approval :)

Consider that in the end the structure here will be very different, I would like to stop passing the hrdw structure all around the modules, that's one of the reasons why I started this.
For the moment we can condense the three lines into one, or we can remove the if and just append the return value, it should not be a problem, you will get a None inside the final list, but we can filter that out.

Or you had a different solution in mind?

**Comment by ErwanAliasr1** [comment]
In fact, I'd love to open a wider discussion around managing parts. I'll open another issue to discuss that.

**Comment by elfosardo** [comment]
> In fact, I'd love to open a wider discussion around managing parts. I'll open another issue to discuss that.

we can also consider this as part of #90

**Comment by elfosardo** [comment]
@dtantsur @tbreeds please take a minute to have a look at this when you get the chance, the idea is to start moving logic out from the main script in modules to be able to stop passing all the data from a function to another.

File: hardware/areca.py [hunk 2]
```diff
@@ -94,60 +96,98 @@ def run_areca(*args):
...
+    _run_areca('set password=0000')
+
+
+def detect():
+    """Detect Areca controller configuration."""
+    hwlist = []
+    device = _sys_info()
+    if not device:
+        sys.stderr.write('Info: No Areca controller found\n')
+        return
```
Comment (dtantsur) [inline hardware/areca.py:154]: Let's do `return []` here and below and get rid of the `if` condition at the caller site?

File: hardware/areca.py [hunk 3]
```diff
@@ -94,60 +96,98 @@ def run_areca(*args):
...
+    if "ControllerName" not in device.keys():
+        sys.stderr.write('Info: Cannot find Areca controller name\n')
+        return
+
+    sys.stderr.write('Info: Areca controller found: %s version %s\n' %
+                     (device['ControllerName'], device['FirmwareVersion']))
+
+    _disable_password()
+
+    for info in device.keys():
```
Comment (dtantsur) [inline hardware/areca.py:164]: nit: we should transform iterating over keys into iterating over items

File: hardware/areca.py (hunk 2 above)
Comment (elfosardo) [reply hardware/areca.py]: done

File: hardware/areca.py (hunk 3 above)
Comment (elfosardo) [reply hardware/areca.py]: yeah, I ended up ding this too

**Comment by elfosardo** [comment]
@ErwanAliasr1 @dtantsur wondering if we can move forward with this now

**Comment by ErwanAliasr1** [comment]
I'm fine with the change but I'd prefer a rebase over a merge please.

**Comment by elfosardo** [comment]
@ErwanAliasr1 rebased :)

**Pull request merged**